File: 2021/julia/src/Day7.jl
module Day7
include("Aoc.jl")
using .Aoc
using Statistics

pos = parse.(Int64, split(Aoc.input_line(7), ","))
tri(n) = n * (n+1) ÷ 2
part1() = (t -> mapreduce(i -> abs(t - i), +, pos))(trunc(Int64, median(pos)))
part2() = findmin(t -> mapreduce(i -> tri(abs(t - i)), +, pos), minimum(pos):maximum(pos))[1]
end

File: 2021/typescript/src/days/day7.ts
import {Day, sum} from "../aoc";

export class Day7 extends Day {
    day = (): number => 7;

    part1 = () => {
        const target = Day7.median(this.pos);
        return this.pos.map(p => Math.abs(p - target)).reduce(sum);
    }

    part2 = () => {
        const min = Math.min(...this.pos);
        return Math.min(...Array.from({length: Math.max(...this.pos) - min})
            .map((v,i) => this.pos.map(p => Day7.triangle(Math.abs(p - i+min))).reduce(sum)));
    }

    private static triangle = (n: number): number =>
        Math.floor(n * (n + 1) / 2);

    private static median = (numbers: number[]): number =>
        numbers.sort((a, b) => a - b)[Math.floor(numbers.length / 2)];

    setup = () => this.pos = this.input[0].split(',').map(Number);

    private pos: number[];
}